Fix win_lgpo execution module

twangboy opened the pull request:
### What does this PR do?
The _processPolicyDefinitions function loads the `adml` files for the corresponding `admx`. The `adml` file usually resides in a subfolder in the `C:\Windows\PolicyDefinitions` directory that is the language code. The default is `en-US`.
When the System Center Operations Manager (SCOM) agent is installed, it adds some additional policy definitions. However, the corresponding `adml` files are not in the standard location (`en-US`). Instead, they are placed in the `en` directory.
This PR will cause the win_lgpo module to check in the `en-US` directory first and then check the `en` directory if still not found.

Additionally, this PR fixes an issue in the `_findOptionValueInSeceditFile` function where the `file.remove` operation would fail due to a lock on the dump file. We don't know where this lock was coming from as python has released the file. Possibly antivirus software.
This adds an additional try/except/else block to catch the error, sleep, and try again. If it fails after 5 tries, it doesn't throw an error, it will just leave the dump file on the system instead of causing the whole module to fail.

@lomeroe Would you mind taking a look? Also, I would like to have your contact information... Could you send me an email at slee@saltstack.com?

### What issues does this PR fix or reference?
Customer Support Issue

### Previous Behavior
```
minion-name:
----------
          ID: configure_windows_update
    Function: lgpo.set
      Result: False
     Comment: An exception occurred in this state: Traceback (most recent call last):
                File "c:\salt\bin\lib\site-packages\salt\state.py", line 1843, in call
                  **cdata['kwargs'])
                File "c:\salt\bin\lib\site-packages\salt\loader.py", line 1795, in wrapper
                  return f(*args, **kwargs)
                File "c:\salt\bin\lib\site-packages\salt\states\win_lgpo.py", line 233, in set_
                  adml_language=adml_language)
                File "c:\salt\bin\lib\site-packages\salt\modules\win_lgpo.py", line 4799, in get_policy_info
                  display_language=adml_language)
                File "c:\salt\bin\lib\site-packages\salt\modules\win_lgpo.py", line 2791, in _processPolicyDefinitions
                  t_admfile))
              SaltInvocationError: An ADML file in the specified ADML language "en-US" and the fallback language "en-US" do not exist for the ADMX "HealthService.admx".
     Started: 13:46:29.740000
    Duration: 172.0 ms
     Changes:
```

### New Behavior
State applied successfully.

### Tests written?
No

twangboy committed:
Fix win_lgpo execution module

The _processPolicyDefinitions fuction loads the adml files for the
corresponding admx. The adml file usually resides in a sub folder in the
C:\Windows\PolicyDefinitions directory that is the language code. The
default is `en-US`.
When the System Center Operations Manager (SCOM) agent is installed, it
adds some addition policy definitions. However, the corresponding adml
files are not in the standard location (`en-US`). Instead they are placed
in the `en` directory.
This PR will cause the win_lgpo module to check in the `en-US` directory
first and then check the `en` directory if still not found.

rallytime commented:
Review(s) requested from @saltstack/team-windows on this one :)

damon-atkins commented:
@lomeroe  is the best person to review

damon-atkins commented:
Only thought I would have is why is the language default hard coded to en-US.  I would have thought it should be what the OS is set too.  I have en-US (lots of files) and en-GB (few files) on my win10.

Need a non US install of windows?

twangboy commented:
@damon-atkins It's not hard-coded to en-US... The default is en-US. If it doesn't find the adml it's looking for then the fallback is hard-coded to en-US.

lomeroe commented:
You *should* be able to add a display_language: en to the state call (display_language=en if directly running the module) to find the right ADMLs (of course, if you weren't en-US the fall back would fail)

Outside of that, the code update looks good and I don't see any reason not to try the language codes without the additional country code.

lomeroe commented:
it appears from a quick glance that pywin32 has the functions to grab the locale, so it could be altered to pull the locale directly from windows by default, but still allow it to be over written by state/module parameter...

lomeroe reviewed (approved):
changes look good to me, the only possible alternative would be to query windows for the current locale initially, but still allow override through state/module parameters

damon-atkins commented:
>>> import locale
>>> locale.getdefaultlocale()
('en_AU', 'cp1252')
>>>locale.windows_locale.get(win32api.GetSystemDefaultLangID())
'en_GB'
>>>locale.windows_locale.get(win32api.GetSystemDefaultLCID())
'en_GB'
>>>locale.windows_locale.get(win32api.GetThreadLocale())
'en_AU'
>>> locale.windows_locale.get(win32api.GetUserDefaultLangID())
'en_AU'
>>> locale.windows_locale.get(win32api.GetUserDefaultLCID())
'en_AU'
locale.windows_locale is a dict.

The user values may change when running as a service under system.

twangboy committed:
Use System Install Language as default fallback

twangboy commented:
@damon-atkins I used `GetSystemDefaultUILanguage` which is supposed to be the Language of the original installation. @lomeroe Is this still good?

lomeroe commented:
@twangboy looks good

damon-atkins commented:
Thought I would add this for info only
>>> locale.windows_locale.get(dll_handle.GetSystemDefaultUILanguage())
'en_GB'

twangboy committed:
Fix problem with file handle

Use with io.open so the handle is realeased automatically....
Use `closed` to detect failure to release the file handle and continue
to close until successful
Use `__opts__` to get `cachedir` instead of __salt__

twangboy commented on salt/modules/win_lgpo.py:
```diff
@@ -2828,13 +2863,14 @@ def _findOptionValueInSeceditFile(option):
+            while not _reader.closed:
+                _reader.close()
             if __salt__['file.file_exists'](_tfile):
                 _ret = __salt__['file.remove'](_tfile)
```
@cachedout Could you look at the above code. On a few Windows boxes the old code here wasn't releasing the file handle before it tried to delete the file. This only happened on 5 machines out of 250 with one of our customers. What do you think of this approach? How best to avoid an infinite loop...?

lomeroe replied:
shouldn't the "with" ensure that it is cleaned up (since the original code did not use a "with")

wouldn't _reader be undefined outside of the "with" as well?

twangboy replied:
You would think, but we tried this on the customer machine and the handle still was not released... adding a `time.sleep(1)` however, allowed enough time for the file handle to be released, but I don't really want to put a 1-second pause in every call to this function as it seems to get called a lot.

That's what I thought initially, but I still have access to `_reader` outside the with statement. I can return `_reader.name` which gives me the full path to the object and `_reader.closed` returns a boolean True/False. None of the calls to `_reader` are throwing an error, which would lead me to believe it's still accessible.

twangboy committed:
Fix INSTALL_LANGUAGE

Replace `_` with `-`
Instead of taking the first 2 elements of the language code to find the
abbreviated language code, do a split off the `-` and take the first
element.

twangboy committed:
Fix some lint

twangboy committed:
Fix syntax error

damon-atkins commented:
time.sleep(float)  supports a fraction.

damon-atkins commented:
Found this
http://forums.whirlpool.net.au/forum-replies.cfm?t=2618245&#r5
Looks like other parts of windows can trap a close, and then do something, e.g. open the file, like a virus scanner.

Looks like we need to add to win_functions a wait_for_rm(file, timeout=5)

damon-atkins commented:
```
def NamedTemporaryFile(mode='w+b', bufsize=-1, suffix="",
                       prefix=template, dir=None, delete=True):
....
# Setting O_TEMPORARY in the flags causes the OS to delete
    # the file when it is closed.  This is only supported by Windows.
    if _os.name == 'nt' and delete:
        flags |= _os.O_TEMPORARY
```

twangboy committed:
Put the file.remove in a try/except/else block

twangboy committed:
Fix typo

twangboy committed:
Fix pylint error

twangboy committed:
Catch CommandExecutionError

twangboy commented:
@cachedout We finally got this fixed. This is ready to merge once tests pass.

twangboy commented:
Rebased...
